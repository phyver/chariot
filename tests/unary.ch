:set unary_constants true
:set continue_on_error true

codata unit where

data nat where Z : unit -> nat
             | S : nat -> nat

codata prod('x,'y) where Fst : prod('x,'y) -> 'x
                       | Snd : prod('x,'y) -> 'y

data list('x) where
    | Nil : unit -> list('x)
    | Cons : prod('x,list('x)) -> list('x)

data oops where Zero : oops
              | Succ : oops -> oops

val map_list f (Nil {}) = Nil {}
  | map_list f (Cons { Fst = x ; Snd = xs }) = Cons { Fst = f x ; Snd = map_list f xs }
:show map_list
