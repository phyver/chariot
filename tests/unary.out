>> option verbose is set to 1
>> option unary_constants is set to true
>> option continue_on_error is set to true
>> coinductive type unit was succesfully defined
>> inductive type nat was succesfully defined
>> coinductive type prod was succesfully defined
>> inductive type list was succesfully defined
** error: constructor Zero is not unary
>> Typing for map_list successful
>> the definition for map_list is complete
>> the definition map_list is provably correct

val
    map_list : ('a → 'b) → list('a) → list('b)
  | map_list f ([] {  }⁰) = [] {  }⁰
  | map_list f (Cons¹ { Fst = x ; Snd = xs }⁰) = Cons¹ { Fst = f x ; Snd = map_list f xs }⁰

